### PR TITLE
Add std.digest.hmac

### PR DESCRIPTION
**Pull request by schuetzm**

### CODE REVIEW

File: std/digest/hmac.d [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    /**
+     * Convenience function for computing an _HMAC digest over a given
+     * piece of data with the specified secret.
+     *
+     * Returns:
+     * The final _HMAC hash.
+     */
+    @nogc
+    auto hmac(scope const(ubyte)[] secret, in ubyte[] data)
```
Comment (WalterBright) [inline std/digest/hmac.d]: ```data``` should be a range, not just an array.

**Comment by WalterBright** [comment]
The interface appears to be, from the unit tests:
```
assert(hmac!SHA1  (key, data).hex == "de7c9b85b8b78aa6bc8a7a36f70a90701c9db4d9");
```
This is not a pipeline interface, nor is ```data``` a range. Should be something like:
```
assert(data.hmac!SHA1(key).hex == "de7c9b85b8b78aa6bc8a7a36f70a90701c9db4d9");
```

File: std/digest/hmac.d [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1)
+            .put(data2);
+    }
+
+    /**
+     * Resets the digest and returns the finished hash.
+     */
+
+    auto finish()
```
Comment (WalterBright) [inline std/digest/hmac.d]: What exactly is this returning? I.e. what is the type? Is it a value, a reference, a range, a GC allocated array, a class?

**Comment by WalterBright** [comment]
How is unittest coverage?
Is there anything that allocates GC memory (other than unittests)?
Can it be made pure and nothrow?

**Comment by WalterBright** [comment]
Also, I'd like to see a ddoc example that is a complete program that reads a file from stdin, writes the digest to stdout, and uses pipelining.

**Comment by schuetzm** [comment]
I reimplemented the helpers following the example of `std.digest.md.md5Of()`.

Coverage: "std/digest/hmac.d is 98% covered"
The one line without coverage is a `map!()` call in one of the unittests, probably because it gets optimized away.

The entire module (without unittests) is `@safe pure nothrow @nogc`. I had originally annotated it as such, but as it's templatized, it wouldn't be possible to use GC-using digest implementations or throwing output ranges, so I removed the annotations. However, I annotated all applicable unittests, so it gets checked.

Also, added the example you requested. The ddoc unittest of `HMAC.put()` already shows the chainability of that method.

**Comment by schuetzm** [comment]
There's a strange failure on the autotester for FreeBSD 64:
```object.Exception@src/rt/minfo.d(162): Aborting: Cycle detected between modules with ctors/dtors:
std.digest.digest -> std.digest.hmac -> std.digest.digest```
None of the modules contain module ctors/dtors, AFAICS.

File: std/digest/hmac.d [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        ubyte[] data1, data2;
+        auto digest = hmac!SHA1("My s3cR3T keY".representation)
+                          .put(data1)
+                          .put(data2)
+                          .finish();
+    }
+
+    /**
+     * Convenience function for computing an _HMAC digest over the given
+     * ranges of data with the specified secret.
```
Comment (WalterBright) [inline std/digest/hmac.d]: ranges? or range?

File: std/digest/hmac.d [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto digest = hmac!SHA1("My s3cR3T keY".representation)
+                          .put(data1)
+                          .put(data2)
+                          .finish();
+    }
+
+    /**
+     * Convenience function for computing an _HMAC digest over the given
```
Comment (WalterBright) [inline std/digest/hmac.d]: s/Convenience function for computing an/Compute an/

File: std/digest/hmac.d [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+     * ranges of data with the specified secret.
+     *
+     * Returns:
+     * The final _HMAC hash.
+     */
+
+    DigestType!H hmac(Range)(auto ref Range range, scope const(ubyte)[] secret)
```
Comment (WalterBright) [inline std/digest/hmac.d]: s/auto ref Range/Range/ should work

File: std/digest/hmac.d [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    if(!isArray!Range && isDigestibleRange!Range)
```
Comment (WalterBright) [inline std/digest/hmac.d]: why can't it be an array?

Comment (schuetzm) [inline std/digest/hmac.d]: Don't know. This is an adaptation of `std.digest.digest.digest()`: https://github.com/D-Programming-Language/phobos/blob/master/std/digest/digest.d#L405 and https://github.com/D-Programming-Language/phobos/blob/master/std/digest/digest.d#L430

I'd have to test whether they can be merged into one function.

**Comment by schuetzm** [comment]
I merged the two helper functions `hmac`.

Unfortunately the resulting function is now `@system` because it needs to pass the address of the `HMAC` local to `std.algorithm.mutation.copy()`, which doesn't take the ouput range by `ref`. A working `scope` implementation will fix this in the future. I could work around it by reimplementing the relevant part of `copy()` locally, if desired (not difficult, only `put()` is used).

File: std/digest/hmac.d [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+ *
+ * This type conforms to $(XREF digest.digest, isDigest).
+ */
+
+version(StdDdoc)
+/// Computes an HMAC over data read from stdin.
+unittest
+{
+    import std.stdio;
+    import std.digest.hmac, std.digest.sha;
```
Comment (andralex) [inline std/digest/hmac.d]: why not all three imports on one line?

File: posix.mak [hunk 8]
```diff
@@ -197,7 +197,7 @@ STD_MODULES = $(addprefix std/, \
   array ascii base64 bigint bitmanip compiler complex concurrency \
   $(addprefix container/, array binaryheap dlist rbtree slist util) \
   conv cstream csv datetime demangle \
-  $(addprefix digest/, digest crc md ripemd sha) \
+  $(addprefix digest/, digest hmac crc md ripemd sha) \
```
Comment (andralex) [inline posix.mak]: sort please

File: std/digest/digest.d [hunk 9]
```diff
@@ -383,7 +384,26 @@ unittest
...
 }
 
-private template isDigestibleRange(Range)
+/**
+ * Checks whether the digest has a $(D blockSize) member, which contains the
+ * digest's internal block size in bits. It is primarily used by $(XREF digest.hmac, HMAC).
+ */
+
+template hasBlockSize(T)
+if(isDigest!T)
```
Comment (andralex) [inline std/digest/digest.d]: space after `if`, `for`, `foreach` etc. throughout - this is phobos style so kinda hard to negotiate against :)

File: std/digest/hmac.d [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+
+License: $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Source: $(PHOBOSSRC std/digest/_hmac.d)
+ */
+
+module std.digest.hmac;
+
+import std.digest.digest : isDigest, hasBlockSize, isDigestibleRange, DigestType;
+import std.traits : isArray;
```
Comment (andralex) [inline std/digest/hmac.d]: this doesn't seem to be used

File: std/digest/hmac.d [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+ * Template API HMAC implementation.
+ *
+ * This implements an _HMAC over the digest H. If H doesn't provide
+ * information about the block size, it can be supplied explicitly using
+ * the second overload.
+ *
+ * This type conforms to $(XREF digest.digest, isDigest).
+ */
+
+version(StdDdoc)
```
Comment (andralex) [inline std/digest/hmac.d]: nice trick

File: std/digest/hmac.d [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+     *
+     * Note:
+     * The constructor leaves the digest in an initialized state, so that this
+     * method only needs to be called if an unfinished digest is to be reused.
+     *
+     * Returns:
+     * A reference to the digest for convenient chaining.
+     */
+
+    ref start() return
```
Comment (andralex) [inline std/digest/hmac.d]: specify the type with the sig?

File: std/digest/hmac.d [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+         .writeln;
+}
+
+template HMAC(H)
+if(isDigest!H && hasBlockSize!H)
+{
+    alias HMAC = HMAC!(H, H.blockSize);
+}
+
+/// ditto
```
Comment (andralex) [inline std/digest/hmac.d]: what's the matching doc here?

File: std/digest/hmac.d [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+
+        start();
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
```
Comment (andralex) [inline std/digest/hmac.d]: This doesn't really test anything and pumps up coverage numbers giving a false sense of security. It should assert that e.g. the contentless digest is equal to whatever appropriate value.

File: std/digest/hmac.d [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1);
+        hmac.start();    // reset digest
+        hmac.put(data2); // start over
+    }
```
Comment (andralex) [inline std/digest/hmac.d]: assert that the result is equal to the appropriate literal

File: std/digest/hmac.d [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+
+        return this;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
```
Comment (andralex) [inline std/digest/hmac.d]: these shouldn't be empty

File: std/digest/hmac.d [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1)
+            .put(data2);
+    }
```
Comment (andralex) [inline std/digest/hmac.d]: make data1 and data2 not empty and verify the result

File: std/digest/hmac.d [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        auto digest = hmac.put(data1)
+                          .put(data2)
+                          .finish();
+    }
```
Comment (andralex) [inline std/digest/hmac.d]: verify that the digest is as expected (again data1 and data2 should not be empty)

File: std/digest/hmac.d [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto digest = hmac!SHA1("My s3cR3T keY".representation)
+                          .put(data1)
+                          .put(data2)
+                          .finish();
+    }
```
Comment (andralex) [inline std/digest/hmac.d]: somebody assert something

File: std/digest/hmac.d (hunk 15 above)
Comment (andralex) [inline std/digest/hmac.d]: assert mumble mumble

File: std/digest/hmac.d [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        import std.algorithm : map;
+        ubyte[] data;
+        auto digest = data
+                      .map!(a => cast(ubyte)(a+1))
+                      .hmac!SHA1("My s3cR3T keY".representation);
+        static assert(is(typeof(digest) == ubyte[20]));
+    }
```
Comment (andralex) [inline std/digest/hmac.d]: also make data not empty and assert something

File: std/digest/sha.d [hunk 21]
```diff
@@ -196,14 +196,16 @@ private ulong rotateRight(ulong x, uint n) @safe pure nothrow @nogc
...
  *
- * The blockSize and digestSize are in bits. However, it's likely easier to
+ * The hashBlockSize and digestSize are in bits. However, it's likely easier to
  * simply use the convenience aliases: SHA1, SHA224, SHA256, SHA384, SHA512,
  * SHA512_224 and SHA512_256.
  *
  * See $(D std.digest.digest) for differences between template and OOP API.
  */
-struct SHA(int blockSize, int digestSize)
+struct SHA(int hashBlockSize, int digestSize)
```
Comment (andralex) [inline std/digest/sha.d]: `uint`?

**Comment by schuetzm** [comment]
@andralex I addressed your comments. The unittests were meant as documentation examples, but I added asserts anyway.

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
@schuetzm See https://issues.dlang.org/show_bug.cgi?id=14517, this is likely the cause.

Basically, if a module imports another module with a template that contains static ctors, and that module instantiates that template, the compiler marks the *importing* module as having static ctors, even if that template's static ctor is not run (because some other module runs it). If you have ideas on how to fix, please add to the bug report.

But in the meantime, I would look for an imported and instantiated template with static ctors/dtors.

**Comment by yebblies** [comment]
Auto-merge toggled off

**Comment by schuetzm** [comment]
@schveiguy Now that I think of it, doesn't that bug only happen with separate compilation, and on all platforms? Phobos didn't use separate compilation when I submitted this (Andrei's change for per-package compilation wasn't in yet), and the failure is restricted to FreeBSD...

**Comment by schveiguy** [comment]
@schuetzm Phobos is separately compiled from druntime. So all you need is one template in druntime that does this, and the errors can happen anywhere. That was the case in the PR that prompted me to file the bug.

**Comment by schveiguy** [comment]
Looks like you found something, now passing freebsd.

**Comment by schuetzm** [comment]
Ok, the culprit is [`isDigestibleRange`](https://github.com/D-Programming-Language/phobos/blob/master/std/digest/digest.d#L386-L396). It imports several other modules, which is a plausible trigger for the bug according to your description.

I tried removing the `package` annotation, moving the imports into the `is`-expression, copying the implementation into `std.digest.hmac` instead of importing it, and turning the constraint of `hmac()` into a `static if` inside the function (and only importing the modules locally). It didn't help. The only thing that works so far is not using it at all.

**Comment by schveiguy** [comment]
What I fear is that there is a druntime module somewhere that has the static ctor, but only for freebsd. This will come up again, we should try and find the culprit and fix it in druntime.

I was fortunate in the original bug report that the PR that triggered the error was *introducing* the static ctor in a template, so it was easy to find. Not sure how I can find it, but I'll try...

**Comment by schveiguy** [comment]
I don't think it's in druntime, did a search for static ctors in there, and none of them are in templates.

I also just realized, phobos unit test build is also done with per-module compilation. So even though phobos may build fine, the unit test build may not.

I'm trying to wrap my head around why FreeBSD only fails. Unfortunately, I don't have a FreeBSD VM set up for testing...

**Comment by schveiguy** [comment]
Found this: https://github.com/schuetzm/phobos/blob/hmac/std/digest/sha.d#L224

I wonder if this is the issue. Is `USE_SSE3` only defined on freebsd? Looking at how it is defined, it's NOT defined if `D_PIC` is defined. So possibly all systems besides FreeBSD define this.

**Comment by schveiguy** [comment]
Note, this IS an issue, even if it's not the culprit here. We should remove that static ctor inside a template. It doesn't look too painful, since the static ctor does not need to be inside the template (it doesn't use any template parameters for its construction).

**Comment by schveiguy** [comment]
See #3408 @schuetzm

**Comment by schuetzm** [comment]
@schveiguy That's it. The tests succeed on the FreeBSDs if I cherry-pick your change.

**Comment by schveiguy** [comment]
blocked until #3408 is pulled

**Comment by schuetzm** [comment]
All green!

**Comment by CyberShadow** [comment]
The Windows makefile changes were insufficient, e.g. the new module's documentation is not built.

An easy way to find all the places where something must be added is to find all occurrences of the name of something else (another Phobos module).

**Comment by schuetzm** [comment]
Did #3543 fix this, or is something still missing?

**Comment by schveiguy** [comment]
#3543 only seems to have minor updates to the table. It doesn't seem to add another file to the generation. See here: http://dtest.thecybershadow.net/results/1dea7ce2c13fecd4211d933c4a6d36a7cef73f59/bdb35215d78ad9f7ad8708cb19b3ca32c7a7cd30/

**Comment by CyberShadow** [comment]
> Did #3543 fix this, or is something still missing?

No, it did not. I added the missing stuff here:

https://github.com/CyberShadow/phobos/commit/7ef38a0e88f3d32c419ac71cf1c315f0c79608eb

> See here: http://dtest.thecybershadow.net/results/1dea7ce2c13fecd4211d933c4a6d36a7cef73f59/bdb35215d78ad9f7ad8708cb19b3ca32c7a7cd30/

That's not relevant in this case because the doc builder only runs on Posix, and that PR only changed Windows makefiles.

**Comment by schveiguy** [comment]
> the doc builder only runs on Posix

Hm... good to know.

**Pull request merged**